### PR TITLE
Add C Shared Library with Python API wrappers

### PR DESCRIPTION
**Pull request by andyrzhao**
The code is structured and named to allow support for other languages besides Python in the future, using a single shared library.

### CODE REVIEW

**Review by arithmetic1728 — DISMISSED** [review]

File: cshared/main.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		copy(cert, pemBytes)
+	}
+	return len(pemBytes)
+}
+
+//export SignForPython
+//
+// SignForPython signs a message digest of length digestLen using a certificate private key
+// specified by configFilePath, storing the result inside a sigHolder byte array of size sigHolderLen.
+func SignForPython(configFilePath *C.char, digest *byte, digestLen int, sigHolder *byte, sigHolderLen int) int {
```
Comment (sai-sunder-s) [inline cshared/main.go:65]: should we check that sigHolderLen is 256? so that we can avoid weird memory issues?

**Review by sai-sunder-s — APPROVED** [review]

File: cshared/main.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+//
+// The shared library exports language-specific wrappers around the Golang
+// client APIs.
+//
+// Example compilation command:
+// go build -buildmode=c-shared -o signer.dylib main.go
+package main
+
+/*
+#include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply cshared/main.go:12]: yes, see an example here: https://karthikkaranth.me/blog/calling-c-code-from-go/

File: cshared/main.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		log.Printf("Could not create client using config %s: %v", configFilePath, err)
+		return nil
+	}
+	defer key.Close()
+
+	certChain := key.CertificateChain()
+	certChainPem := []byte{}
+	for i := 0; i < len(certChain); i++ {
+		certPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certChain[i]})
+		certChainPem = append(certChainPem, certPem...)
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply cshared/main.go:39]: yes, cert chain is cert piled together, something like:
```
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
```

File: cshared/main.go (hunk 1 above)
Comment (arithmetic1728) [reply cshared/main.go:65]: when it is used in google-auth lib, I am passing in sigHolder buffer of length 2000, which is more than enough. This function will fill the signature into the buffer, and return the actual length of the signature. I made a suggestion to add a size check.

File: cshared/main.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	} else {
+		signature, signErr = key.Sign(nil, digestSlice, crypto.SHA256)
+	}
+	if signErr != nil {
+		log.Printf("failed to sign hash: %v", signErr)
+		return 0
+	}
+
+	// Create a Go buffer around the output buffer and copy the signature into the buffer
+	outBytes := unsafe.Slice(sigHolder, sigHolderLen)
```
Comment (arithmetic1728) [inline cshared/main.go:110]:
```suggestion
	outBytes := unsafe.Slice(sigHolder, sigHolderLen)
	if sigHolderLen < len(signature) {
	    log.Print("The sigHolder buffer size %d is smaller than the signature size %d", sigHolderLen, len(signature))
	    return 0
	}
```

**Review by arithmetic1728 — DISMISSED** [review]

File: cshared/main.go (hunk 1 above)
Comment (andyrzhao) [reply cshared/main.go:65]: Applied your suggestion. Thanks!

**Review by sai-sunder-s — APPROVED** [review]

**Pull request merged**